# Pandas–продвинутый

### Формулировка задания:

Необходимо загрузить и обработать предложенный датасет (Титаник), провести аналитику данных с помощью фреймворка Python Pandas.

### Планируемый результат:

1. Датасет загружен в Colab
2. В датасете отсутствуют пустые ячейки
3. Представлена простая аналитика по датасету

### Описания плана работы:

1. Создать новый ноутбук в Colab
2. Сохранить датасет в Google диске и загрузить в ноутбук Colab https://www.kaggle.com/datasets/yasserh/titanic-dataset/code
*Загрузить датасет по ссылке из интернета (репозитория Kaggle)
3. Перечень и описание столбцов:
  Survived выжил (1 - да, 0 - нет)
  Pclass Класс круиза
  Name ФИО
  Sex Пол
  Age Возраст
  SibSp числобратьев, сестер или супругов на борту у человека
  Parch количество родителей или детей, с которыми путешествовал каждый пассажир
  Ticket Номер билета
  Fare Цена билета
  Cabin Номер каюты
  Embarked Порт посадки
4. Определить количество пустых ячеек
5. *Заполнить пустые ячейки и спользуя любую логику замещения данных
6. Если пункт 5 невыполнен, то удалить строки имеющие пустые ячейки.
7. По каждому признаку произвести аналитику:
  1. Количество уникальных значений
  2. Минимальное и максимальное значение
8. *С помощью аналитики определить влияние всех признаков на признак Survived (выживание).

### Пример:
Parch и SibSP отрицательно влияли навыживание при крушении, так как чем больше эти параметры тем ниже процент выживания (одиночке выжить проще).
Результатом домашнего задания будет Таблица в Colab с комментариями в качестве ответов на пункты ДЗ.

### Перечень инструментов, необходимых для реализации деятельности:
1) Google Colab https://colab.research.google.com/

2) *PyCharm

In [8]:
#Устанавливаем библиотеки Pandas и NumPy
!pip3 install pandas

In [43]:
#Устанавливаем библиотеку matplotlib
!pip3 install matplotlib

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 622.2 kB/s eta 0:00:00m eta 0:00:010:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 300.7/300.7 KB 583.4 kB/s eta 0:00:00 kB/s eta 0:00:01:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 969.9 kB/s eta 0:00:00m eta 0:00:010:01:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 861.5 kB/s eta 0:00:00m eta 0:00:010:01:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 759.2 kB/s eta 0:00:00m eta 0:00:010:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.3/98.3 KB 1.2 MB/s eta 0:00:00 MB/s eta 0:00:01:01


In [45]:
#Устанавливаем библиотеку seaborn
!pip3 install seaborn

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 293.3/293.3 KB 921.9 kB/s eta 0:00:00 kB/s eta 0:00:01:01


In [46]:
# Подгружаем библиотеки Pandas, NumPy
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [32]:
#Загружаем датасет
df = pd.read_csv('Titanic-Dataset.csv', delimiter=',', index_col=[0])
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 891 entries, 1 to 891
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Name      891 non-null    object 
 3   Sex       891 non-null    object 
 4   Age       714 non-null    float64
 5   SibSp     891 non-null    int64  
 6   Parch     891 non-null    int64  
 7   Ticket    891 non-null    object 
 8   Fare      891 non-null    float64
 9   Cabin     204 non-null    object 
 10  Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 83.5+ KB


Исходя из описательной статистики предполагаем что количество записей в датасете 891
видим что в столбцах с возрастом, номером кабины и порт посадки есть пропуски.
видим что в датасете 3 типа данных float64, int64, object
размер датасета узнаем с помощью функции shape,
и статистические данные по числовым значениям узнаем с помощью функции describe

In [26]:
count = df.shape #кортеж из количества строк и столбцов
print(f'количество строк - {count[0]}, количество столбцов - {count[1]}')

количество строк - 891, количество столбцов - 11


In [27]:
df.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [36]:
df.isnull().sum()

Survived    0
Pclass      0
Name        0
Sex         0
Age         0
SibSp       0
Parch       0
Ticket      0
Fare        0
Cabin       0
Embarked    2
dtype: int64

Заполним пропуски в числовых столбцах с возрастом и номер кабины (логично было бы заполнить номер кабины по
имеющим номерам кабин по данной фамилии имеющих родственников, SibSp числобратьев, сестер или супругов на борту
у человека и Parch количество родителей или детей, с которыми путешествовал каждый пассажир), в excel я бы разбил по столбцам столбец Name и искал там родственников, будем думать что датасет составлен так что родственники идут друг за другом и для заполнения номера кабины используем функцию 

In [33]:
df['Cabin'] = df['Cabin'].fillna(method='bfill')
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   C85        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250  C123        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   E46        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   B42        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500  C148        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

In [34]:
df['Cabin'] = df['Cabin'].fillna(method='ffill')
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   C85        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250  C123        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   E46        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   B42        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500  C148        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500  C148        Q  

[891 rows x 11 columns]

In [35]:
median_age = df.Age.median()
df.fillna({'Age' : median_age}, inplace = True)
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female  28.0   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   C85        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250  C123        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   E46        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   B42        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500  C148        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500  C148        Q  

[891 rows x 11 columns]

In [37]:
# для практики можно удалить 2 строки Embarked Порт посадки 
df.dropna(axis = 'index', subset = ['Embarked'], inplace = True)
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female  28.0   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   C85        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250  C123        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   E46        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   B42        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500  C148        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500  C148        Q  

[889 rows x 11 columns]

In [40]:
#Выводим количество уникальных значений по столбцам
df.nunique()

Survived      2
Pclass        3
Name        889
Sex           2
Age          88
SibSp         7
Parch         7
Ticket      680
Fare        247
Cabin       147
Embarked      3
dtype: int64

In [41]:
# Выводим данные по максимальным минимальным значениям числовых столбцов
df.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  889.000000  889.000000  889.000000  889.000000  889.000000  889.000000
mean     0.382452    2.311586   29.315152    0.524184    0.382452   32.096681
std      0.486260    0.834700   12.984932    1.103705    0.806761   49.697504
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   22.000000    0.000000    0.000000    7.895800
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   35.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [49]:
df_part_class = df[['Pclass', 'Survived']]
fuel_type_rate = df_part_class.groupby('Pclass').apply(lambda x: (len(x[x['Survived'] == 1])/len(x), len(x[x['Survived'] == 1])))


for i, rate in fuel_type_rate.items():
  print(f'с классом круиза {i} доля выживших: {rate[0]}, общее число : {rate[1]}')


с классом круиза 1 доля выживших: 0.6261682242990654, общее число : 134
с классом круиза 2 доля выживших: 0.47282608695652173, общее число : 87
с классом круиза 3 доля выживших: 0.24236252545824846, общее число : 119


вывод: чем выше класс билета пасажира, тем больше выживших

In [51]:
df_part_age = df[['Age', 'Survived']]
fuel_type_rate = df_part_age.groupby('Age').apply(lambda x: (len(x[x['Survived'] == 1])/len(x), len(x[x['Survived'] == 1])))


for i, rate in fuel_type_rate.items():
  print(f'с возрастом {i} доля выживших: {rate[0]}, общее число : {rate[1]}')

с возрастом 0.42 доля выживших: 1.0, общее число : 1
с возрастом 0.67 доля выживших: 1.0, общее число : 1
с возрастом 0.75 доля выживших: 1.0, общее число : 2
с возрастом 0.83 доля выживших: 1.0, общее число : 2
с возрастом 0.92 доля выживших: 1.0, общее число : 1
с возрастом 1.0 доля выживших: 0.7142857142857143, общее число : 5
с возрастом 2.0 доля выживших: 0.3, общее число : 3
с возрастом 3.0 доля выживших: 0.8333333333333334, общее число : 5
с возрастом 4.0 доля выживших: 0.7, общее число : 7
с возрастом 5.0 доля выживших: 1.0, общее число : 4
с возрастом 6.0 доля выживших: 0.6666666666666666, общее число : 2
с возрастом 7.0 доля выживших: 0.3333333333333333, общее число : 1
с возрастом 8.0 доля выживших: 0.5, общее число : 2
с возрастом 9.0 доля выживших: 0.25, общее число : 2
с возрастом 10.0 доля выживших: 0.0, общее число : 0
с возрастом 11.0 доля выживших: 0.25, общее число : 1
с возрастом 12.0 доля выживших: 1.0, общее число : 1
с возрастом 13.0 доля выживших: 1.0, общее чис

Вывод: младенцы с возрастом до 2 лет большая доля выживших, далее кореляция по возрасту проподает

In [53]:
df_part_sex = df[['Sex', 'Survived']]
fuel_type_rate = df_part_sex.groupby('Sex').apply(lambda x: (len(x[x['Survived'] == 1])/len(x), len(x[x['Survived'] == 1])))


for i, rate in fuel_type_rate.items():
  print(f'с полом {i} доля выживших: {rate[0]}, общее число : {rate[1]}')

с полом female доля выживших: 0.7403846153846154, общее число : 231
с полом male доля выживших: 0.18890814558058924, общее число : 109


Вывод: доля выживших женщин больше чем мужчин.